### PR TITLE
rev to the latest protocol

### PR DESCRIPTION
**Pull request by devoncarew**
- rev to the latest protocol
- this introduces the ability to evaluate in the scope of another object
- support parsing the new `map<string,string>` type
- fix a todo around code gen for ElementLists

@danrubel

### CODE REVIEW

File: dart/tool/dart/generate_dart.dart [hunk 1]
```diff
@@ -1113,10 +1125,23 @@ class MethodParser extends Parser {
...
+          ref.arrayDepth++;
+        }
+      } else if (peek().text == '<') {
+        // handle generics
+        expect('<');
+        ref.genericTypes = [];
+        while (peek().text != '>') {
+          Token genericTypeName = expectName();
+          ref.genericTypes.add(new TypeRef(_coerceRefType(genericTypeName.text)));
+          consume(',');
```
Comment (danrubel) [inline dart/tool/dart/generate_dart.dart]: consume --> expect   ?

File: dart/tool/java/generate_java.dart [hunk 2]
```diff
@@ -610,9 +645,22 @@ class MethodParser extends Parser {
...
+        }
+      } else if (peek().text == '<') {
+        // handle generics
+        expect('<');
+        ref.genericTypes = [];
+        while (peek().text != '>') {
+          Token genericTypeName = expectName();
+          ref.genericTypes
+              .add(new TypeRef(_coerceRefType(genericTypeName.text)));
+          consume(',');
```
Comment (danrubel) [inline dart/tool/java/generate_java.dart]: Should consume be changed to expect ?

File: dart/tool/dart/generate_dart.dart (hunk 1 above)
Comment (devoncarew) [inline dart/tool/dart/generate_dart.dart]: I think the commas are optional here, in that `a, b, c` - c won't have a trailing comma.

Comment (devoncarew) [inline dart/tool/dart/generate_dart.dart]: I could change this to break out of the loop if we don't find a comma -

Comment (danrubel) [inline dart/tool/dart/generate_dart.dart]: Good point and I don't think it needs to be changed.

**Comment by danrubel** [comment]
Still LGTM

**Pull request merged**